### PR TITLE
refactor(debugging): use new probe status endpoint

### PR DESCRIPTION
**Pull request by P403n1x87**
We refactor the upload logic to redirect the probe status messages to the new dedicated endpoint on the agent. The upload is done via a multipart POST request with the event.json file carrying the same JSON payload that was being uploaded along with log signals with in the previous implementation.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

**Review by emmettbutler — APPROVED** [review]

File: ddtrace/debugging/_encoding.py [hunk 1]
```diff
@@ -74,7 +71,7 @@ def put(self, item: Any) -> int:
         """Enqueue the given item and returns its encoded size."""
 
     @abc.abstractmethod
-    def encode(self) -> Optional[bytes]:
+    def flush(self) -> Optional[bytes]:
         """Encode the given item."""
```
Comment (shatzi) [inline ddtrace/debugging/_encoding.py:75]: change comment

File: ddtrace/debugging/_probe/status.py [hunk 2]
```diff
@@ -56,39 +72,43 @@ def _payload(
...
+                else:
+                    meter.increment("upload.success")
+                    meter.distribution("upload.size", len(payload))
+        except Exception:
+            log.error("Failed to write payload", exc_info=True)
+            meter.increment("error")
 
+    def _write(self, probe: Probe, status: str, message: str, error: Optional[ErrorInfo] = None) -> None:
+        try:
+            self._write_payload_with_backoff(
```
Comment (shatzi) [inline ddtrace/debugging/_probe/status.py:97]: we are sending single message per probe status?

**Review by shatzi — APPROVED** [review]

File: ddtrace/debugging/_uploader.py [hunk 3]
```diff
@@ -84,9 +84,9 @@ def upload(self) -> None:
...
     def periodic(self) -> None:
         """Upload the buffer content to the logs intake."""
-        count = self._encoder.count
+        count = self._queue.count
         if count:
-            payload = self._encoder.encode()
+            payload = self._queue.flush()
             if payload is not None:
                 try:
                     self._write_with_backoff(payload)
```
Comment (shatzi) [inline ddtrace/debugging/_uploader.py:92]: I find that the implementation of _uploader and _probeStatusLogger while using similar terms and design have different details that make it hard to see. for example _probeStatusLogger use _write_payload_with_backoff that is called inside the flush while this class use flush to only generate payload and then call _write_with_backoff (different name).

I don't think this need to be on this PR but maybe we can look into simplify those two classes?

File: ddtrace/debugging/_probe/status.py [hunk 4]
```diff
@@ -56,53 +73,68 @@ def _payload(
...
+            msgs.append(self._queue.get_nowait())
+
+        try:
+            self._write_payload_with_backoff(
+                multipart(
+                    parts=[
+                        FormData(
+                            name="event",
+                            filename="event.json",
+                            data=f"[{','.join(msgs)}]",
```
Comment (shatzi) [inline ddtrace/debugging/_probe/status.py:116]: are there limit on number of messages we send? as this is not log backend anymore... how big the messages can be?

**Review by shatzi — APPROVED** [review]

File: ddtrace/debugging/_probe/status.py (hunk 4 above)
Comment (P403n1x87) [reply ddtrace/debugging/_probe/status.py:116]: On the assumption that we have a hard limit on probes, I don't think we ever risk of getting near the size limit (which I'd expect to be ~ 1 MB).

**Pull request merged**